pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,56 @@
+package server
+
+import (
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type tagService struct {
+	distribution.TagService
+
+	imageStream *imageapi.ImageStream
+}
+
+func (t tagService) Get(ctx context.Context, tag string) (distribution.Descriptor, error) {
+	context.GetLogger(ctx).Debug("(tagService).Get: starting")
+
+	te := imageapi.LatestTaggedImage(t.imageStream, tag)
+	if te == nil {
+		return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+	}
+	dgst, err := digest.ParseDigest(te.Image)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	return distribution.Descriptor{Digest: dgst}, nil
+}
+
+func (t tagService) All(ctx context.Context) ([]string, error) {
+	context.GetLogger(ctx).Debug("(tagService).All: starting")
+
+	tags := []string{}
+	for tag := range t.imageStream.Status.Tags {
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, digest distribution.Descriptor) ([]string, error) {
+	context.GetLogger(ctx).Debug("(tagService).Lookup: starting")
+	return t.TagService.Lookup(ctx, digest)
+}
+
+// TODO(legion): Rewrite this method to store tag in etcd.
```

Comment:
> The Get/Lookup and Tag are necessary for sure. We'll need the migration in the future if we don't implement the other methods as well. Otherwise we'll have tag entries in the storage.

We already have tags on a filesystem. If you create a tag using the `docker tag`, it will be created by the `docker/distribution` on the filesystem. If you're going to use `oc tag`, they will be in the etcd.

In this aspect, nothing has changed.

> I don't understand why do we need migration controller now if we implement them here.

Tags already created by `docker` needs to be imported into etcd.